### PR TITLE
Expand performance tests comparison & output

### PR DESCRIPTION
**Pull request by swissspidy**
1. Run the performance test results for the target branch (if a PR) or previous commit (if on trunk)
2. Create a nice little comparison table for the two results
3. Leave a comment on the PR with the results ([example](https://github.com/swissspidy/wordpress-develop/pull/33#issuecomment-1679084103))
    Screenshot 2023-08-15 at 17 17 39
4. Add results to the GitHub workflow summary ([example](https://github.com/swissspidy/wordpress-develop/actions/runs/5868414189?pr=33))
    Screenshot 2023-08-15 at 17 17 57

To-do:

* [x] Set up a GitHub access token on `WordPress/wordpress-develop` for writing comments on PRs

Trac ticket: 

See:

* https://core.trac.wordpress.org/ticket/58358
* https://core.trac.wordpress.org/ticket/56150
* https://core.trac.wordpress.org/ticket/58359

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by swissspidy** [comment]
PR against my repo for testing: https://github.com/swissspidy/wordpress-develop/pull/33

**Comment by swissspidy** [comment]
Just realized that this is PR number 5000 😄  That's _alot_ of PRs.

File: .github/workflows/performance.yml
Comment (mukeshpanchal27) [inline .github/workflows/performance.yml]:
```suggestion
  # - Add workflow summary.
  # - Check if a comment was already made.
  # - Comment on PR with test results.
```

Add missing end period.

File: .github/workflows/performance.yml
Comment (mukeshpanchal27) [inline .github/workflows/performance.yml]: Is it common within the condition? Could we relocate it to the bottom following the condition?

File: .github/workflows/performance.yml
Comment (mukeshpanchal27) [inline .github/workflows/performance.yml]:
```suggestion
        run: |
          if [[ -z "$TARGET_REF" ]]; then
            git fetch -n origin $TARGET_SHA
```
Please add indentation for improved visibility.

File: tests/performance/compare-results.js
Comment (mukeshpanchal27) [inline tests/performance/compare-results.js]:
```suggestion
// The previous commit's results.
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Thank you, @swissspidy, for the pull request. It looks good to me overall. I've provided some feedback points for your consideration.

Since #4999 has been committed, could you please incorporate the metrics `TTFB`, `LCP`, and `LCP - TTFB`?

**Comment by swissspidy** [comment]
> Since #4999 has been committed, could you please incorporate the metrics `TTFB`, `LCP`, and `LCP - TTFB`?

That should already be the case now that I merged the latest changes from trunk. You can see the results here: https://github.com/swissspidy/wordpress-develop/pull/33#issuecomment-1679084103

File: tests/performance/compare-results.js
Comment (mukeshpanchal27) [inline tests/performance/compare-results.js]:
```suggestion
// The current commit's results.
```

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks @swissspidy for the update. The changes look good to me.

**Comment by oandregal** [comment]
> Leave a comment on the PR with the results (https://github.com/swissspidy/wordpress-develop/pull/33#issuecomment-1679084103)

Would you be open to leave this for a follow-up, so we can merge the rest of it ASAP. This PR provides an enormous value, so the faster we merge, the better.

**Comment by swissspidy** [comment]
Yes, definitely open to that.

File: .github/workflows/performance.yml
Comment (felixarntz) [inline .github/workflows/performance.yml]: Per my feedback on Slack in the performance chat, can we pull that out into a separate ticket and PR? While it's technically certainly good, the metrics are too unstable at this point to post them _that obviously_ on every PR. People that aren't familiar with the metrics work we've been doing (which certainly is the majority of contributors) may over-interpret a slightly positive or slightly negative result, not knowing that it just comes from variance.

In the interest of time, I think we should merge the rest and reconsider this part for a Future Release.
```suggestion
```

**Review by felixarntz — CHANGES_REQUESTED** [review]
@swissspidy Looks great to me, just the one thing related to @oandregal's comment I think we should remove for now.

**Comment by swissspidy** [comment]
Removed the PR comment part now.

I think it looks good, but I'll wait for a final review from @desrosj

File: tests/performance/compare-results.js
Comment (desrosj) [inline tests/performance/compare-results.js]: Could we link the SHA to the corresponding commit on GitHub? Not a blocker, but would be nice to have the ability to click this to view the changeset.

File: tests/performance/compare-results.js
Comment (desrosj) [inline tests/performance/compare-results.js]: Same here for linking the SHA. Not a blocker, but would be nice to be able to click through to the commit.

File: tests/performance/compare-results.js
Comment (desrosj) [inline tests/performance/compare-results.js]:
```suggestion
let summaryMarkdown = `# Performance Test Results\n\n`;
```

File: tests/performance/compare-results.js
Comment (desrosj) [inline tests/performance/compare-results.js]:
```suggestion
	summaryMarkdown += `🛎️ Performance test results are in!\n\n`;
```

File: tests/performance/compare-results.js
Comment (desrosj) [inline tests/performance/compare-results.js]:
```suggestion
	summaryMarkdown += `🛎️ Performance test results for ${ process.env.GITHUB_SHA } are in!\n\n`;
```

Comment (desrosj) [reply tests/performance/compare-results.js]: Personal preference, but I feel the emoji at the beginning breaks up the section a bit.

File: tests/performance/compare-results.js
Comment (desrosj) [inline tests/performance/compare-results.js]:
```suggestion
	summaryMarkdown += `## ${ title }\n\n`;
```

File: .github/workflows/performance.yml
Comment (desrosj) [inline .github/workflows/performance.yml]: Can we move these `env` variables to the workflow level?

`$TARGET_SHA` is used within another step further down in the workflow. Defining it once at the top would avoid having to repeat these and also gives a better spot to include any inline documentation we'd like it have explaining their values.

File: .github/workflows/performance.yml
Comment (desrosj) [inline .github/workflows/performance.yml:185]: Would it be helpful to define this as a script in the `package.json` file? Not sure if being able to say `npm run test:performance-results` would be useful. But it would definitely make moving or changing the command easier down the road having to only change it in one place.

File: tests/performance/compare-results.js
Comment (desrosj) [inline tests/performance/compare-results.js:72]: Could we just add a brief docblock here?

**Review by desrosj — APPROVED** [review]
Looks good! Thanks all!

I left mainly minor nit picks mostly around improving the formatting of the summary and minor adjustments to the workflow itself. I don't really consider any of them hard blockers, though.

File: tests/performance/compare-results.js
Comment (swissspidy) [reply tests/performance/compare-results.js]: Yeah sure. In the GitHub PR it's auto-linked, but in the summary unfortunately not 😞

File: .github/workflows/performance.yml
Comment (swissspidy) [reply .github/workflows/performance.yml:185]: Probably best to do that in a future step.

Right now we have 3 different scripts:

* `results.js` - print results for an individual run
* `compare-results.js` - compare results from 2 runs and print a result (console + markdown)
* `log-results.js` - send results to dashboard

I envision a future where `results.js` is absorbed in the test runner, i.e. running tests automatically prints nice results at the end.

File: tests/performance/compare-results.js
Comment (desrosj) [inline tests/performance/compare-results.js]: @swissspidy I think this is reversed. The text should be within `[]` and the URL within `()`.

Comment (swissspidy) [reply tests/performance/compare-results.js]: Yep, just noticed as well

**Comment by swissspidy** [comment]
Links are now fixed: https://github.com/WordPress/wordpress-develop/actions/runs/6051491344?pr=5000

**Review by felixarntz — APPROVED** [review]

**Comment by swissspidy** [comment]
Committed in https://core.trac.wordpress.org/changeset/56506

**Pull request closed without merging**
